reference/reference.go

```diff
@@ -0,0 +1,181 @@
+}
+
+// Normalize returns a reference in its normalized form, meaning repo name
+// will not contain default hostname nor library/ prefix for official images.
+func Normalize(ref Named) Named {
+	host, remoteName := SplitHostname(ref)
+	if host == DefaultHostname {
+		if strings.HasPrefix(remoteName.Name(), DefaultRepoPrefix) {
+			return withNewName(ref, strings.TrimPrefix(remoteName.Name(), DefaultRepoPrefix))
+		}
+		return remoteName
+	}
+	return ref
+}
+
+// WithHostname returns a referece with full hostname and index path.
+func WithHostname(ref Named) Named {
+	host, remoteName := SplitHostname(ref)
+	return withNewName(ref, host+"/"+remoteName.Name())
+}
+
+// SplitHostname splits a named reference into a
+// hostname and name string. If no valid hostname is
+// found, the hostname is empty and the full value
+// is returned as name
+func SplitHostname(named Named) (string, Named) {
+	indexName, remoteName := distreference.SplitHostname(named)
+	if indexName == "" || (!strings.ContainsAny(indexName, ".:") && indexName != "localhost") {
+		// This is a Docker Index repos (ex: samalba/hipache or ubuntu)
+		// 'docker.io'
+		indexName = DefaultHostname
+		remoteName = named.Name()
+	}
+	if indexName == LegacyDefaultHostname {
+		indexName = DefaultHostname
+	}
+	if indexName == DefaultHostname && !strings.Contains(remoteName, "/") {
+		remoteName = DefaultRepoPrefix + remoteName
+	}
+	return indexName, withNewName(named, remoteName)
+}
+
+// WithName returns a named object representing the given string. If the input
+// is invalid ErrReferenceInvalidFormat will be returned.
+func WithName(name string) (Named, error) {
+	if err := validateName(name); err != nil {
+		return nil, err
+	}
+	return distreference.WithName(name)
+}
+
+// WithTag combines the name from "name" and the tag from "tag" to form a
+// reference incorporating both the name and the tag.
+func WithTag(name Named, tag string) (NamedTagged, error) {
+	return distreference.WithTag(name, tag)
+}
+
+// WithDigest combines the name from "name" and the digest from "digest" to form
+// a reference incorporating both the name and the digest.
+func WithDigest(name Named, digest digest.Digest) (Canonical, error) {
+	return distreference.WithDigest(name, digest)
+}
+
+// WithDefaultTag adds a default tag to a reference if it only has a repo name.
+func WithDefaultTag(ref Named) Named {
+	if IsNameOnly(ref) {
+		ref, _ = WithTag(ref, DefaultTag)
+	}
+	return ref
+}
+
+// IsNameOnly returns true is reference only contains a repo name.
+func IsNameOnly(ref Named) bool {
+	if _, ok := ref.(NamedTagged); ok {
+		return false
+	}
+	if _, ok := ref.(Canonical); ok {
+		return false
+	}
+	return true
+}
+
+func withNewName(ref Named, name string) Named {
+	newRef, err := WithName(name)
+	if err != nil {
+		return ref
+	}
+	if digested, isCanonical := ref.(Canonical); isCanonical {
+		newRef, err = WithDigest(newRef, digested.Digest())
+		if err != nil {
+			return ref
```

Comment:
I think because they are interfaces we should not fully rely on that. `Digest()` can return anything if the original named ref was not properly parsed or it may even be another object that just satisfies the interface.